Source/Diagnostics/ReducedDiags/ColliderRelevant.cpp

```diff
@@ -0,0 +1,559 @@
+#include "Utils/WarpXConst.H"
+#include "Utils/TextMsg.H"
+#include "WarpX.H"
+
+#include <AMReX_Algorithm.H>
+#include <AMReX_Array.H>
+#include <AMReX_Array4.H>
+#include <AMReX_Box.H>
+#include <AMReX_Dim3.H>
+#include <AMReX_Extension.H>
+#include <AMReX_FArrayBox.H>
+#include <AMReX_FabArray.H>
+#include <AMReX_GpuQualifiers.H>
+#include <AMReX_IndexType.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_PODVector.H>
+#include <AMReX_ParIter.H>
+#include <AMReX_ParallelDescriptor.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_ParticleReduce.H>
+#include <AMReX_Particles.H>
+#include <AMReX_Print.H>
+#include <AMReX_REAL.H>
+#include <AMReX_Reduce.H>
+#include <AMReX_Tuple.H>
+#include <AMReX_Vector.H>
+
+#include <ablastr/coarsen/sample.H>
+#include <ablastr/warn_manager/WarnManager.H>
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <fstream>
+#include <limits>
+#include <map>
+#include <vector>
+
+using namespace amrex;
+
+ColliderRelevant::ColliderRelevant (std::string rd_name)
+: ReducedDiags{rd_name}
+{
+    // read colliding species names - must be 2
+    amrex::ParmParse pp_rd_name(rd_name);
+    pp_rd_name.getarr("species", m_beam_name);
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        m_beam_name.size() == 2u,
+        "Collider-relevant diagnostics must involve exactly two species");
+
+    // RZ coordinate is not working
+#if (defined WARPX_DIM_RZ)
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(false,
+        "Collider-relevant diagnostics do not work in RZ geometry.");
+#endif
+
+    ablastr::warn_manager::WMRecordWarning(
+        "DIAGNOSTICS",
+        "The collider-relevant reduced diagnostic is meant for \
+        colliding species propagating along the z direction.",
+        ablastr::warn_manager::WarnPriority::low);
+
+    ablastr::warn_manager::WMRecordWarning(
+        "DIAGNOSTICS",
+        "The collider-relevant reduced diagnostic only considers the \
+        coarsest level of refinement for the calculations involving chi.",
+        ablastr::warn_manager::WarnPriority::low);
+
+    // get WarpX class object
+    auto& warpx = WarpX::GetInstance();
+
+    // get MultiParticleContainer class object
+    const MultiParticleContainer& mypc =  warpx.GetPartContainer();
+
+    // get species names (std::vector<std::string>)
+    const std::vector<std::string> species_names = mypc.GetSpeciesNames();
+
+    // loop over species
+    for (int i_s = 0; i_s < 2; ++i_s)
+    {
+        // get WarpXParticleContainer class object
+        const WarpXParticleContainer& myspc = mypc.GetParticleContainerFromName(m_beam_name[i_s]);
+
+        const bool is_photon = myspc.AmIA<PhysicalSpecies::photon>();
+
+        // photon number density is not available yet
+        WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+            !is_photon,
+            "Collider-relevant diagnostic does not work for colliding photons yet");
+    }
+
+    auto all_diag_names = std::vector<std::string>{};
```

Comment:
This block deserves an inline comment :sweat_smile: 
_What does it do?_ (Not _how_.)